services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/monitor/dynamic_sidecar_api.py

```diff
@@ -0,0 +1,180 @@
+import logging
+import traceback
+from typing import Any, Dict, Optional
+
+import httpx
+from fastapi import FastAPI
+
+from ....core.settings import DynamicSidecarSettings
+from ..exceptions import MonitorException
+from .models import MonitorData
+
+logger = logging.getLogger(__name__)
+
+
+KEY_DYNAMIC_SIDECAR_API_CLIENT = f"{__name__}.DynamicSidecarClient"
+
+
+def get_url(dynamic_sidecar_endpoint: str, postfix: str) -> str:
+    """formats and returns an url for the request"""
+    url = f"{dynamic_sidecar_endpoint}{postfix}"
+    return url
+
+
+def log_httpx_http_error(url: str, method: str, formatted_traceback: str) -> None:
+    # mainly used to debug issues with the API
+    logging.debug(
+        (
+            "%s -> %s generated:\n %s\nThe above logs can safely "
+            "be ignored, except when debugging an issue "
+            "regarding the dynamic-sidecar"
+        ),
+        method,
+        url,
+        formatted_traceback,
+    )
+
+
+class DynamicSidecarClient:
+    """Will handle connections to the service sidecar"""
+
+    def __init__(self, app: FastAPI):
+        self._app = app
+        self._heatlth_request_timeout = httpx.Timeout(1.0, connect=1.0)
+
+        dynamic_sidecar_settings: DynamicSidecarSettings = (
+            app.state.settings.dynamic_services.dynamic_sidecar
+        )
+
+        self.httpx_client = httpx.AsyncClient(
+            timeout=httpx.Timeout(
+                dynamic_sidecar_settings.api_request_timeout,
+                connect=1.0,
+            )
+        )
+
+    async def close(self):
+        await self.httpx_client.aclose()
+
+    async def is_healthy(self, dynamic_sidecar_endpoint: str) -> bool:
+        """retruns True if service is UP and running else False"""
+        url = get_url(dynamic_sidecar_endpoint, "/health")
+        try:
+            # this request uses a very short timeout
+            response = await self.httpx_client.get(
+                url=url, timeout=self._heatlth_request_timeout
+            )
+            if response.status_code != 200:
+                return False
+
+            return response.json()["is_healthy"]
+        except httpx.HTTPError:
+            return False
+
+    async def containers_inspect(
+        self, dynamic_sidecar_endpoint: str
+    ) -> Optional[Dict[str, Any]]:
+        """returns: None in case of error, otherwise a dict will be returned"""
```

Comment:
actually it would be more reliable to raise exceptions no? I mean if someone uses this code without checking too much the errors will be swallowed.

Reply:
Normally I would agree with you. But it makes no sense to raise an exception when the scope of this function is not to have any errors where it is called. You just end up moving the error catching.

Since this is an interface used by the `monitor` I will not change it and keep it as it is.

Reply:
well on the other hand if someone else than you changes the monitor. he/she will not know that the exceptions are not returned and will be confused that the errors are swallowed. how does None means error? how do you distinguish from the client whether it's an actual error or that the container does not exist?